File: blog2/internal/entity/file.go
package entity

import (
	"regexp"
	"time"
)

type FileID string

type FileType string

const (
	FileTypeImage   FileType = "image"
	FileTypeVideo   FileType = "video"
	FileTypeText    FileType = "text"
	FileTypeUnknown FileType = "unknown"
)

func NewFileTypeFromMimeType(mimeType string) FileType {
	// image
	reMimeTypeImage := regexp.MustCompile("image/*")
	if reMimeTypeImage.MatchString(mimeType) {
		return FileTypeImage
	}
	// video
	reMimeTypeVideo := regexp.MustCompile("video/*")
	if reMimeTypeVideo.MatchString(mimeType) {
		return FileTypeVideo
	}
	// text
	reMimeTypeText := regexp.MustCompile("text/*")
	if reMimeTypeText.MatchString(mimeType) {
		return FileTypeText
	}
	return FileTypeUnknown
}

type File struct {
	ID        FileID    `json:"id"`
	Type      FileType  `json:"type"`
	MediaType string    `json:"mediaType"` // Ex) application/json; charset=utf8
	CreatedAt time.Time `json:"createdAt"`
	UpdatedAt time.Time `json:"updatedAt"`
}
type FileThumbnailID string

type FileThumbnail struct {
	ID        FileThumbnailID `json:"id"`
	FileID    FileID          `json:"fileId"`
	MediaType string          `json:"mediaType"` // Ex) application/json; charset=utf8
}

type FileAndThumbnail struct {
	File          *File          `json:"file"`
	FileThumbnail *FileThumbnail `json:"fileThumbnail"`
}

File: blog2/_ts/src/entity/file.ts

export type FileId = string;
export type FileType = 'image' | 'video' | 'text' | 'unknown';
export interface File {
    id: FileId,
    type: FileType,
    mediaType: string,
    createdAt: string,
    updatedAt: string,
}
export type FileThumbnailId = string;
export interface FileThumbnail {
    id: FileThumbnailId,
    fileId: FileId,
    mediaType: string,
}
export interface FileAndThumbnail {
    file: File;
    fileThumbnail: FileThumbnail | null;
}

export function getThumbnailUrl(a: FileThumbnail | null): string {
    const baseUrl = window.env.baseUrlFileThumbnail;
    if (a === null) {
        return "";
    }
    if (baseUrl.endsWith("/")) {
        return `${baseUrl}${a.id}`;
    }
    return `${baseUrl}/${a.id}`;
}

export function getFileUrl(a: File): string {
    const baseUrl = window.env.baseUrlFile;
    if (baseUrl.endsWith("/")) {
        return `${baseUrl}${a.id}`;
    }
    return `${baseUrl}/${a.id}`;
}

export function generateFileMarkdownURL(file: File): string {
    const url = getFileUrl(file)
    if (file.type === "image") {
        return `![](${url})`;
    }
    return `[${file.id}](${url})`;
}